### ISSUE TITLE
Missing `cbits/categorise.cpp`

### ISSUE THREAD

**Issue by silky** [flora-pm/flora-server#131]
## Environment

* Version or commit: `6c552d737e6b36c0acec9cffa94a3d74b50f6f64`

## Expected Behaviour

The build to work.

## Actual Behaviour

```
[nix-shell:~/dev/ext/flora-server]$ make build
Build profile: -w ghc-8.10.7 -O1
In order, the following will be built (use -v for more details):
 - flora-0.0.1.0 (lib) (first run)
 - flora-0.0.1.0 (test:flora-test) (first run)
 - flora-0.0.1.0 (exe:flora-server) (first run)
 - flora-0.0.1.0 (exe:flora-cli) (first run)
Preprocessing library for flora-0.0.1.0..
Building library for flora-0.0.1.0..
<command line>: does not exist: cbits/categorise.cpp
```

## Instructions for reproduction

```
> nix-shell
> make build
```

**Comment by tchoutri** [comment]
@silky Do you use Soufflé Datalog v2.2 ?

**Comment by silky** [comment]
No ... ? I'm not sure.

(I didn't do anything, I just used the nix shell)

**Comment by tchoutri** [comment]
@silky could you give me the result of `souffle --version` in the nix shell ?

**Comment by silky** [comment]
Hmm...

```
[nix-shell:~/dev/ext/flora-server]$ souffle --version
/nix/store/4mb403c7vxwhf0a0hsp0b0i7rk9v7qmk-souffle-2.2/bin/souffle: /nix/store/02hgizi6v46z0a1giis0q93wq6g22d77-ncurses-6.2/lib/libncursesw.so.6: no version information available (required by /nix/store/4mb403c7vxwhf0a0hsp0b0i7rk9v7qmk-souffle-2.2/bin/souffle)
/nix/store/4mb403c7vxwhf0a0hsp0b0i7rk9v7qmk-souffle-2.2/bin/souffle: /nix/store/02hgizi6v46z0a1giis0q93wq6g22d77-ncurses-6.2/lib/libncursesw.so.6: no version information available (required by /nix/store/4mb403c7vxwhf0a0hsp0b0i7rk9v7qmk-souffle-2.2/bin/souffle)
Souffle: (32bit Domains)
Copyright (c) 2016-19 The Souffle Developers.
Copyright (c) 2013-16 Oracle and/or its affiliates.
```

```
[nix-shell:~/dev/ext/flora-server]$ which souffle
/nix/store/4mb403c7vxwhf0a0hsp0b0i7rk9v7qmk-souffle-2.2/bin/souffle
```

**Comment by silky** [comment]
@tchoutri "success"

The build works now, but I get some moderately distressing "errors" that appear to just be warnings that are ignored.

```
cbits/categorise.cpp: In member function ‘bool souffle::t_btree_iii__0_1_2__111::t_comparator_0::less(const t_tuple&, const t_tuple&) const’:

cbits/categorise.cpp:17:260: error:
     warning: suggest parentheses around ‘&&’ within ‘||’ [-Wparentheses]
       17 |   return (ramBitCast<RamSigned>(a[0]) < ramBitCast<RamSigned>(b[0]))|| (ramBitCast<RamSigned>(a[0]) == ramBitCast<RamSigned>(b[0])) && ((ramBitCast<RamSigned>(a[1]) < ramBitCast<RamSigned>(b[1]))|| (ramBitCast<RamSigned>(a[1]) == ramBitCast<RamSigned>(b[1])) && ((ramBitCast<RamSigned>(a[2]) < ramBitCast<RamSigned>(b[2]))));
          |                                                                                                                                                                                                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   |
17 |   return (ramBitCast<RamSigned>(a[0]) < ramBitCast<RamSigned>(b[0]))|| (ramBitCast<RamSigned>(a[0]) == ramBitCast<RamSigned>(b[0])) && ((ramBitCast<RamSigned>(a[1]) < ramBitCast<RamSigned>(b[1]))|| (ramBitCast<RamSigned>(a[1]) == ramBitCast<RamSigned>(b[1])) && ((ramBitCast<RamSigned>(a[2]) < ramBitCast<RamSigned>(b[2]))));
   |                                                                                                                                                                                                                                                                    ^

cbits/categorise.cpp:17:133: error:
     warning: suggest parentheses around ‘&&’ within ‘||’ [-Wparentheses]
       17 |   return (ramBitCast<RamSigned>(a[0]) < ramBitCast<RamSigned>(b[0]))|| (ramBitCast<RamSigned>(a[0]) == ramBitCast<RamSigned>(b[0])) && ((ramBitCast<RamSigned>(a[1]) < ramBitCast<RamSigned>(b[1]))|| (ramBitCast<RamSigned>(a[1]) == ramBitCast<RamSigned>(b[1])) && ((ramBitCast<RamSigned>(a[2]) < ramBitCast<RamSigned>(b[2]))));
          |                                                                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   |
17 |   return (ramBitCast<RamSigned>(a[0]) < ramBitCast<RamSigned>(b[0]))|| (ramBitCast<RamSigned>(a[0]) == ramBitCast<RamSigned>(b[0])) && ((ramBitCast<RamSigned>(a[1]) < ramBitCast<RamSigned>(b[1]))|| (ramBitCast<RamSigned>(a[1]) == ramBitCast<RamSigned>(b[1])) && ((ramBitCast<RamSigned>(a[2]) < ramBitCast<RamSigned>(b[2]))));
   |                                                                                                                                     ^
cbits/categorise.cpp: In member function ‘bool souffle::t_btree_ii__0_1__11__10::t_comparator_0::less(const t_tuple&, const t_tuple&) const’:

cbits/categorise.cpp:120:133: error:
     warning: suggest parentheses around ‘&&’ within ‘||’ [-Wparentheses]
      120 |   return (ramBitCast<RamSigned>(a[0]) < ramBitCast<RamSigned>(b[0]))|| (ramBitCast<RamSigned>(a[0]) == ramBitCast<RamSigned>(b[0])) && ((ramBitCast<RamSigned>(a[1]) < ramBitCast<RamSigned>(b[1])));
          |                                                                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    |
120 |   return (ramBitCast<RamSigned>(a[0]) < ramBitCast<RamSigned>(b[0]))|| (ramBitCast<RamSigned>(a[0]) == ramBitCast<RamSigned>(b[0])) && ((ramBitCast<RamSigned>(a[1]) < ramBitCast<RamSigned>(b[1])));
    |                                                                                                                                     ^
cbits/categorise.cpp: In member function ‘bool souffle::t_btree_iii__0_1_2__111::t_comparator_0::less(const t_tuple&, const t_tuple&) const’:

cbits/categorise.cpp:17:260: error:
     warning: suggest parentheses around ‘&&’ within ‘||’ [-Wparentheses]
       17 |   return (ramBitCast<RamSigned>(a[0]) < ramBitCast<RamSigned>(b[0]))|| (ramBitCast<RamSigned>(a[0]) == ramBitCast<RamSigned>(b[0])) && ((ramBitCast<RamSigned>(a[1]) < ramBitCast<RamSigned>(b[1]))|| (ramBitCast<RamSigned>(a[1]) == ramBitCast<RamSigned>(b[1])) && ((ramBitCast<RamSigned>(a[2]) < ramBitCast<RamSigned>(b[2]))));
          |                                                                                                                                                                                                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   |
17 |   return (ramBitCast<RamSigned>(a[0]) < ramBitCast<RamSigned>(b[0]))|| (ramBitCast<RamSigned>(a[0]) == ramBitCast<RamSigned>(b[0])) && ((ramBitCast<RamSigned>(a[1]) < ramBitCast<RamSigned>(b[1]))|| (ramBitCast<RamSigned>(a[1]) == ramBitCast<RamSigned>(b[1])) && ((ramBitCast<RamSigned>(a[2]) < ramBitCast<RamSigned>(b[2]))));
   |                                                                                                                                                                                                                                                                    ^

cbits/categorise.cpp:17:133: error:
     warning: suggest parentheses around ‘&&’ within ‘||’ [-Wparentheses]
       17 |   return (ramBitCast<RamSigned>(a[0]) < ramBitCast<RamSigned>(b[0]))|| (ramBitCast<RamSigned>(a[0]) == ramBitCast<RamSigned>(b[0])) && ((ramBitCast<RamSigned>(a[1]) < ramBitCast<RamSigned>(b[1]))|| (ramBitCast<RamSigned>(a[1]) == ramBitCast<RamSigned>(b[1])) && ((ramBitCast<RamSigned>(a[2]) < ramBitCast<RamSigned>(b[2]))));
          |                                                                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   |
17 |   return (ramBitCast<RamSigned>(a[0]) < ramBitCast<RamSigned>(b[0]))|| (ramBitCast<RamSigned>(a[0]) == ramBitCast<RamSigned>(b[0])) && ((ramBitCast<RamSigned>(a[1]) < ramBitCast<RamSigned>(b[1]))|| (ramBitCast<RamSigned>(a[1]) == ramBitCast<RamSigned>(b[1])) && ((ramBitCast<RamSigned>(a[2]) < ramBitCast<RamSigned>(b[2]))));
   |                                                                                                                                     ^
cbits/categorise.cpp: In member function ‘bool souffle::t_btree_ii__0_1__11__10::t_comparator_0::less(const t_tuple&, const t_tuple&) const’:

cbits/categorise.cpp:120:133: error:
     warning: suggest parentheses around ‘&&’ within ‘||’ [-Wparentheses]
      120 |   return (ramBitCast<RamSigned>(a[0]) < ramBitCast<RamSigned>(b[0]))|| (ramBitCast<RamSigned>(a[0]) == ramBitCast<RamSigned>(b[0])) && ((ramBitCast<RamSigned>(a[1]) < ramBitCast<RamSigned>(b[1])));
          |                                                                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    |
120 |   return (ramBitCast<RamSigned>(a[0]) < ramBitCast<RamSigned>(b[0]))|| (ramBitCast<RamSigned>(a[0]) == ramBitCast<RamSigned>(b[0])) && ((ramBitCast<RamSigned>(a[1]) < ramBitCast<RamSigned>(b[1])));
```

**Comment by tchoutri** [comment]
@silky Yes this is normal, they will vanish once we migrate to using Soufflé 2.3, thanks to @luc-tielen

**Comment by luc-tielen** [comment]
> The build works now, but I get some moderately distressing "errors" that appear to just be warnings that are ignored.

@silky these are "harmless" errors. In this case the logic is correct (souffle is a well-tested compiler), though it might be hard to read and determine the precedence of the `&&` and `||` operators in the code that it generated. In a PR I made that was merged into souffle 2.3, the extra parentheses were added. :smiley:
